CQM-39

MatiXOfficial opened the pull request:
Dynamic sample and CDNs settings change (backend)

MatiXOfficial committed:
Added update-parameters mapping

MatiXOfficial committed:
Scheduler delay is now updated with update-parameters

MatiXOfficial committed:
Added cdns to update-parameters and some small changes.

- Improved updating parameters
- Ensured thread-safety

MatiXOfficial committed:
updated README

MatiXOfficial committed:
Replaced fixedDelay with fixedRate in the scheduler

MatiXOfficial committed:
Added logging for update-parameters

MatiXOfficial committed:
Replaced PostMapping with PutMapping

MatiXOfficial committed:
Updated ThroughputService to use updated parameters

MatiXOfficial committed:
Improved updateParameters

MatiXOfficial committed:
Fixed activeTestsIntentiy in updateParameters and updated README

MatiXOfficial committed:
.gitignore update

MatiXOfficial committed:
Integreated monitor parameters with the database

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/controller/MonitoringController.java:
```diff
@@ -70,4 +70,19 @@ public AllParametersResponseDTO getAll(
     }
+
+    @PutMapping("/update-parameters")
+    public void ParametersDTO(
```
rename to something like updateParamenters, always camelCase on methods

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/data/dto/ConfigParametersDTO.java:
```diff
@@ -0,0 +1,18 @@
+
+    private List<String> cdns;
+
+    private int activeSamplingRate;
```
It's better to use `Integer` for DTO fields, as it allows handling null values

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/data/dto/ConfigParametersDTO.java:
```diff
@@ -0,0 +1,18 @@
+
+    private List<String> cdns;
+
+    private int activeSamplingRate;
+    private int activeTestIntensity;
+
+    private int passiveSamplingRate;
```
Let's add some very basic validation for those fields, at least `@NotNull` and `@Min(1)`

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/data/model/ConfigCdn.java:
```diff
@@ -0,0 +1,21 @@
+public class ConfigCdn {
+
+    @Id
+    @Column(nullable = false, updatable = false)
```
String columns should have `length` parameter specified, otherwise they are stored as `TEXT` type in the database

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/data/model/ConfigCdn.java:
```diff
@@ -0,0 +1,21 @@
+@Entity
+@Builder
+@AllArgsConstructor
+@Table
```
not needed

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/data/model/ConfigCdn.java:
```diff
@@ -0,0 +1,21 @@
+@Table
+public class ConfigCdn {
+
+    @Id
```
It's a good practice to use numeric IDs everywhere

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/data/model/ConfigSample.java:
```diff
@@ -0,0 +1,36 @@
+@Builder
+@AllArgsConstructor
+@Table
+@EqualsAndHashCode
```
included in `@Data` annotation

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/data/model/ConfigSample.java:
```diff
@@ -0,0 +1,36 @@
+@Entity
+@Builder
+@AllArgsConstructor
+@Table
```
not needed

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/data/model/ConfigSample.java:
```diff
@@ -0,0 +1,36 @@
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    @EqualsAndHashCode.Exclude
+    @ToString.Exclude
```
any reason for that?

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/data/model/ConfigSample.java:
```diff
@@ -0,0 +1,36 @@
+    private long id;
+
+    @Column(nullable = false, updatable = false)
+    @EqualsAndHashCode.Exclude
+    @ToString.Exclude
```
same as above

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/data/repository/ConfigSampleRepository.java:
```diff
@@ -0,0 +1,13 @@
+
+    ConfigSample findFirstByOrderByIdDesc();
+
+    ConfigSample findFirstByTimestampGreaterThanEqual(Instant date);
```
date => timestamp

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/service/SchedulerService.java:
```diff
@@ -1,27 +1,34 @@
+    public void schedulePing() {
+        if (schedulePingCounter == 0) {
+            pingService.doMeasurement();
+            schedulePingCounter = configSampleRepository.findFirstByOrderByIdDesc().getActiveSamplingRate();
```
it might not be guaranteed, that the newest id is also the largest. Better sort by timestamp

HubertBalcerzak replied:
configuration should be accessed through a more readable method from `ParameterService`

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/service/UpdateParametersService.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.util.List;
+
+public interface UpdateParametersService {
```
rename to ParameterService

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/service/ThroughputServiceImpl.java:
```diff
@@ -26,31 +24,30 @@
+                                 ConfigSampleRepository configSampleRepository) throws PcapNativeException {
+        this.configuration = configuration;
         this.dataRepository = dataRepository;
+        this.configCdnRepository = configCdnRepository;
+        this.configSampleRepository = configSampleRepository;
```
as above, everything should be exposed by `ParameterService`

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/service/UpdateParametersServiceImpl.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    @Override
+    public void updateCdns(List<String> cdns) {
```
add `@Transactional` annotation

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/service/UpdateParametersServiceImpl.java:
```diff
@@ -0,0 +1,110 @@
+
+    @Override
+    public void updateCdns(List<String> cdns) {
+        if (cdns != null) {
```
if we add `@NotNull` to the dto, this null check becomes unnecessary

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/service/UpdateParametersServiceImpl.java:
```diff
@@ -0,0 +1,110 @@
+        if (cdns != null) {
+            configCdnRepository.deleteAll();
+            for (String cdn : cdns) {
+                configCdnRepository.save(ConfigCdn.builder().cdn(cdn).build());
```
With an object this simple, just using the constructor would be fine

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/service/UpdateParametersServiceImpl.java:
```diff
@@ -0,0 +1,110 @@
+    @Override
+    public void updateSampleParameters(int activeSamplingRate, int activeTestIntensity, int passiveSamplingRate) {
+
+        ConfigSample prevParameters = configSampleRepository.findFirstByOrderByIdDesc();
```
order by timestamp, can be replaced with a service method, once it exists

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/service/UpdateParametersServiceImpl.java:
```diff
@@ -0,0 +1,110 @@
+        ConfigSample prevParameters = configSampleRepository.findFirstByOrderByIdDesc();
+        ConfigSample configSample = new ConfigSample();
+
+        if (activeSamplingRate > 0) {
+            configSample.setActiveSamplingRate(activeSamplingRate);
+        } else {
+            configSample.setActiveSamplingRate(prevParameters.getActiveSamplingRate());
+        }
+
+        if (activeTestIntensity > 0) {
+            configSample.setActiveTestIntensity(activeTestIntensity);
+        } else {
+            configSample.setActiveTestIntensity(prevParameters.getActiveTestIntensity());
+        }
+
+        if (passiveSamplingRate > 0) {
+            configSample.setPassiveSamplingRate(passiveSamplingRate);
+        } else {
+            configSample.setPassiveSamplingRate(prevParameters.getPassiveSamplingRate());
+        }
+
+        if (!prevParameters.equals(configSample)) {
+            configSample.setTimestamp(Instant.now());
+            configSampleRepository.save(configSample);
+
+            logger.info("Updated the sample parameters: " + configSample);
+        } else {
+            logger.info("None of the sample parameters were updated.");
+        }
```
validation at the DTO level will be sufficent

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/controller/MonitoringController.java:
```diff
@@ -70,4 +70,19 @@ public AllParametersResponseDTO getAll(
         );
     }
+
+    @PutMapping("/update-parameters")
+    public void ParametersDTO(
+            @Valid ConfigParametersDTO configParametersDTO
+    ) {
+        List<String> cdns = configParametersDTO.getCdns();
+        updateParametersService.updateCdns(cdns);
+
+        int activeSamplingRate = configParametersDTO.getActiveSamplingRate();
+        int activeTestIntensity = configParametersDTO.getActiveTestIntensity();
+
+        int passiveSamplingRate = configParametersDTO.getPassiveSamplingRate();
+
+        updateParametersService.updateSampleParameters(activeSamplingRate, activeTestIntensity, passiveSamplingRate);
+    }
```
speaking in REST terms, this endpoint refers to a different resource (parameters) and should be moved to a different controller, with different base path like `/api/parameters`

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/controller/MonitoringController.java:
```diff
@@ -70,4 +70,19 @@ public AllParametersResponseDTO getAll(
         );
     }
+
+    @PutMapping("/update-parameters")
```
updating a REST resource should look something like `PUT /api/path/to/resource`, so in this case `PUT /api/parameters`

HubertBalcerzak commented on spring-app/src/main/java/pl/edu/agh/cqm/controller/MonitoringController.java:
```diff
@@ -70,4 +70,19 @@ public AllParametersResponseDTO getAll(
+
+        updateParametersService.updateSampleParameters(activeSamplingRate, activeTestIntensity, passiveSamplingRate);
+    }
 }
```
endpoint for getting current parameters is missing. It will be needed for displaying settings form in react app

MatiXOfficial committed:
Many small changes

- some refactoring
- better DTOs validation
- timestamp instead of id in finding the latest monitor config

MatiXOfficial committed:
Moved parameter getters to ParameterService

MatiXOfficial committed:
Improved endpoints for the parameters

- new path
- new Get endpoint for parameters
- improved getCdns method in ParametersService

MatiXOfficial committed:
Updated README

MatiXOfficial committed:
Added @AllArgsConstructor

MatiXOfficial committed:
Final tweaks

- Added RequiredArgsConstructor

MatiXOfficial committed:
Merge branch 'master' into CQM-39

# Conflicts:
#	spring-app/src/main/java/pl/edu/agh/cqm/configuration/CqmConfiguration.java
#	spring-app/src/main/java/pl/edu/agh/cqm/service/PingServiceImpl.java

MatiXOfficial committed:
Merge with master
